*** cugraph/python/pylibcugraph/setup.py-dist	2021-09-24 08:27:59.252681765 +0900
--- cugraph/python/pylibcugraph/setup.py	2021-09-24 08:28:41.963467351 +0900
***************
*** 130,136 ****
                    cuda_lib_dir,
                    os.path.join(os.sys.prefix, "lib")
                ],
!               libraries=['cudart', 'cusparse', 'cusolver', 'cugraph', 'nccl'],
                language='c++',
                extra_compile_args=['-std=c++17'])
  ]
--- 130,136 ----
                    cuda_lib_dir,
                    os.path.join(os.sys.prefix, "lib")
                ],
!               libraries=['cudart', 'cusparse', 'cusolver', 'cugraph', 'nccl', 'spdlog', 'fmt'],
                language='c++',
                extra_compile_args=['-std=c++17'])
  ]
***************
*** 149,155 ****
            # "Operating System :: OS Independent",
            "Programming Language :: Python",
            "Programming Language :: Python :: 3.6",
!           "Programming Language :: Python :: 3.7"
        ],
        # Include the separately-compiled shared library
        author="NVIDIA Corporation",
--- 149,157 ----
            # "Operating System :: OS Independent",
            "Programming Language :: Python",
            "Programming Language :: Python :: 3.6",
!           "Programming Language :: Python :: 3.7",
!           "Programming Language :: Python :: 3.8",
!           "Programming Language :: Python :: 3.9"
        ],
        # Include the separately-compiled shared library
        author="NVIDIA Corporation",
*** cugraph/python/cugraph/setup.py-dist	2021-09-24 08:36:06.075011328 +0900
--- cugraph/python/cugraph/setup.py	2021-09-24 08:36:47.899120694 +0900
***************
*** 111,117 ****
                    cuda_lib_dir,
                    os.path.join(os.sys.prefix, "lib")
                ],
!               libraries=['cudart', 'cusparse', 'cusolver', 'cugraph', 'nccl'],
                language='c++',
                extra_compile_args=['-std=c++17'])
  ]
--- 111,117 ----
                    cuda_lib_dir,
                    os.path.join(os.sys.prefix, "lib")
                ],
!               libraries=['cudart', 'cusparse', 'cusolver', 'cugraph', 'nccl', 'spdlog', 'fmt'],
                language='c++',
                extra_compile_args=['-std=c++17'])
  ]
***************
*** 193,199 ****
            # "Operating System :: OS Independent",
            "Programming Language :: Python",
            "Programming Language :: Python :: 3.6",
!           "Programming Language :: Python :: 3.7"
        ],
        # Include the separately-compiled shared library
        author="NVIDIA Corporation",
--- 193,201 ----
            # "Operating System :: OS Independent",
            "Programming Language :: Python",
            "Programming Language :: Python :: 3.6",
!           "Programming Language :: Python :: 3.7",
!           "Programming Language :: Python :: 3.8",
!           "Programming Language :: Python :: 3.9"
        ],
        # Include the separately-compiled shared library
        author="NVIDIA Corporation",
