*** rmm/cmake/Modules/RMM_thirdparty.cmake-dist	2021-09-08 00:28:50.770329991 +0900
--- rmm/cmake/Modules/RMM_thirdparty.cmake	2021-09-08 00:30:00.851565025 +0900
***************
*** 1,6 ****
  # spdlog
  
! set(RMM_MIN_VERSION_spdlog 1.7.0)
  
  if(NOT TARGET spdlog)
    CPMFindPackage(
--- 1,6 ----
  # spdlog
  
! set(RMM_MIN_VERSION_spdlog 1.9.2)
  
  if(NOT TARGET spdlog)
    CPMFindPackage(
***************
*** 24,31 ****
  CPMFindPackage(
    NAME Thrust
    GITHUB_REPOSITORY NVIDIA/thrust
!   GIT_TAG 1.12.0
!   VERSION 1.12.0
    GIT_SHALLOW TRUE
    # If there is no pre-installed thrust we can use, we'll install our fetched copy together with RMM
    OPTIONS "THRUST_INSTALL TRUE")
--- 24,31 ----
  CPMFindPackage(
    NAME Thrust
    GITHUB_REPOSITORY NVIDIA/thrust
!   GIT_TAG 1.14.0
!   VERSION 1.14.0
    GIT_SHALLOW TRUE
    # If there is no pre-installed thrust we can use, we'll install our fetched copy together with RMM
    OPTIONS "THRUST_INSTALL TRUE")
