*** cudf/cpp/cmake/thirdparty/get_thrust.cmake-dist	2021-09-08 00:34:44.016595987 +0900
--- cudf/cpp/cmake/thirdparty/get_thrust.cmake	2021-09-08 00:35:12.397103047 +0900
***************
*** 62,67 ****
      rapids_export_find_package_root(INSTALL Thrust [=[${CMAKE_CURRENT_LIST_DIR}/../../../include/libcudf/Thrust/]=] cudf-exports)
  endfunction()
  
! set(CUDF_MIN_VERSION_Thrust 1.12.0)
  
  find_and_configure_thrust(${CUDF_MIN_VERSION_Thrust})
--- 62,67 ----
      rapids_export_find_package_root(INSTALL Thrust [=[${CMAKE_CURRENT_LIST_DIR}/../../../include/libcudf/Thrust/]=] cudf-exports)
  endfunction()
  
! set(CUDF_MIN_VERSION_Thrust 1.14.0)
  
  find_and_configure_thrust(${CUDF_MIN_VERSION_Thrust})
