*** faiss/CMakeLists.txt-dist	2021-07-26 22:15:11.832996139 +0900
--- faiss/CMakeLists.txt	2021-07-26 22:15:28.179946655 +0900
***************
*** 7,13 ****
  cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
  
  project(faiss
!   VERSION 1.6.4
    DESCRIPTION "A library for efficient similarity search and clustering of dense vectors."
    HOMEPAGE_URL "https://github.com/facebookresearch/faiss"
    LANGUAGES CXX)
--- 7,13 ----
  cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
  
  project(faiss
!   VERSION 1.7.1
    DESCRIPTION "A library for efficient similarity search and clustering of dense vectors."
    HOMEPAGE_URL "https://github.com/facebookresearch/faiss"
    LANGUAGES CXX)
